src/libraries/Microsoft.Extensions.Options/tests/Microsoft.Extensions.Options.Tests/OptionsMonitorTest.cs

```diff
@@ -418,5 +419,51 @@ public ChangeTokenSource(IChangeToken changeToken)
 
             public IChangeToken GetChangeToken() => _changeToken;
         }
+
+        [Fact]
+        public void CallsPublicGetOrAddForCustomOptionsCache()
+        {
+            DerivedOptionsCache derivedOptionsCache = new();
+            CreateMonitor(derivedOptionsCache).Get(null);
+            Assert.Equal(1, derivedOptionsCache.GetOrAddCalls);
+
+            ImplementedOptionsCache implementedOptionsCache = new();
+            CreateMonitor(implementedOptionsCache).Get(null);
+            Assert.Equal(1, implementedOptionsCache.GetOrAddCalls);
+
+            static OptionsMonitor<FakeOptions> CreateMonitor(IOptionsMonitorCache<FakeOptions> cache) =>
+                new OptionsMonitor<FakeOptions>(
+                    new OptionsFactory<FakeOptions>(Enumerable.Empty<IConfigureOptions<FakeOptions>>(), Enumerable.Empty<IPostConfigureOptions<FakeOptions>>()),
+                    Enumerable.Empty<IOptionsChangeTokenSource<FakeOptions>>(),
+                    cache);
+        }
+
+        private class DerivedOptionsCache : OptionsCache<FakeOptions>
+        {
+            public int GetOrAddCalls { get; private set; }
+
+            public override FakeOptions GetOrAdd(string? name, Func<FakeOptions> createOptions)
+            {
+                GetOrAddCalls++;
+                return base.GetOrAdd(name, createOptions);
+            }
+        }
+
+        private class ImplementedOptionsCache : IOptionsMonitorCache<FakeOptions>
+        {
+            public int GetOrAddCalls { get; private set; }
+
+            public void Clear() => throw new NotImplementedException();
+
+            public FakeOptions GetOrAdd(string? name, Func<FakeOptions> createOptions)
+            {
+                GetOrAddCalls++;
+                return createOptions();
+            }
+
+            public bool TryAdd(string? name, FakeOptions options) => throw new NotImplementedException();
+
+            public bool TryRemove(string? name) => throw new NotImplementedException();
+        }
```

Comment:
Are there standard patterns for testing the behavior that a cached Get() doesn't allocate?

Reply:
I doubt there is a commonly used behavior for this.
But if this is important I would suggest that within a NoGC region you measure the allocated bytes before the get call and after the get call (https://github.com/dotnet/runtime/blob/59b5585bd29333a8f00f3819f908930705c484f4/src/coreclr/System.Private.CoreLib/src/System/GC.cs#L356-L357).

### Example
```cs
void TestMethod() {
    Assert.True(GC.TryStartNoGCRegion());

    long allocatedBytesBefore = GC.GetAllocatedBytesForCurrentThread();
    // my cool operation (Get)
    long allocatedBytesAfter = GC.GetAllocatedBytesForCurrentThread();
    Assert.Equal(allocatedBytesBefore, allocatedBytesAfter);

    GC.EndNoGCRegion();
}
```

How reliable this allocated bytes count is I can't tell you though. Probably someone from the GC area would have to tell you.

/cc @Maoni0

Reply:
This seems like a good approach assuming that the allocated byte count is reliable. 

I think we'd also want to layer on something like https://tsuyoshiushio.medium.com/controlling-the-serial-and-parallel-test-on-xunit-6174326da196 to ensure that the test is not parallelized with other tests which could eat up the no-GC region.

Reply:
Actually, wouldn't it make sense to put an AssertExtension for the allocations into the TestUtilities? 
Is there anything against it? I could imagine that this will be a test pattern more often in the future.

```cs
Assert.NoAllocations(static () => _ = monitor.CurrentValue);
```

Reply:
@deeprobin this seems straightforward to add to https://github.com/dotnet/runtime/blob/main/src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs if you think that's useful. I don't know what the guidelines are for adding new methods there.

In the snippet you show I don't think we could use `static` because you're capturing `monitor` in the closure. However, I don't think that matters since the method would check for allocations when calling the provided `Action`, not from creating it.

Anyway, let me know if you'd like me to move this to be a TestExtensions helper!

Reply:
> In the snippet you show I don't think we could use static because you're capturing monitor in the closure. However, I don't think that matters since the method would check for allocations when calling the provided Action, not from creating it.

You're right.

> this seems straightforward to add to https://github.com/dotnet/runtime/blob/main/src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs if you think that's useful. I don't know what the guidelines are for adding new methods there.

I think there are hardly any guidelines (except the style guidelines).

> Anyway, let me know if you'd like me to move this to be a TestExtensions helper!

In my opinion, of course, that would make sense but I might wait for an okay from a maintainer of the repository.
@stephentoub What do you think?

Reply:
I think it would be OK to add an AssertExtension for this functionality. It would be useful in other places. It doesn't have to happen in this PR though. It could happen the next time we need it.

Reply:
> It could happen the next time we need it.

Makes sense. Rule of three and all that.